### PR TITLE
fix(webview): send stable view-state id on launch and re-pin per-view state

### PR DESCRIPTION
**Pull request by easonLiangWorldedtech**
Part of the vps2 durable per-view state series — tracked in easonLiangWorldedtech/Zoo-Code#41 (cross-repo: standalone a+d measured against the stack base; the displayed vs-main diff includes lower units until they merge).

Issue (created at PR-open time): https://github.com/Zoo-Code-Org/Zoo-Code/issues/1551

## What

Persists each webview's stable state identity at launch and makes launch-time per-view state re-pin correctly. The webview now carries a `viewStateId` (created and persisted via the webview state API, with an in-memory fallback) that is posted with `webviewDidLaunch` and persisted on the provider via `setViewStateId`, re-keying the pre-launch temporary state entry to the launching webview. When the view-local API profile is invalid at launch, the view is re-pinned to the still-valid shared global selection with a view-local write only — the shared global is repaired only when its own selection is also invalid. `updateSettings` is routed through `provider.setValue` so view-local buffer/pin sync stays consistent with the other mutation paths.

## Design decisions

- The view-state id is owned by the webview (`VSCodeAPIWrapper.getViewStateId`): it reuses the id persisted in webview state, creates one (`crypto.randomUUID`, with a timestamp+random fallback) and persists it via `setState`; when storage is unavailable it falls back to an in-memory field. The id is best-effort — the launch message carries `viewStateId: undefined` when the helper is unavailable, and the provider degrades to the shared-global path.
- Re-pin is view-local: `provider.saveViewState("currentApiConfigName", name)` writes the view's buffer/pin without touching the shared global selection; the legacy global repair (global write + `activateProviderProfile`) runs only when the shared global selection is also invalid.
- Validation order on launch: merged (view-local) name first, then the shared global selection, then the first listed profile — matching the merged `getState()` semantics from F1b.
- `updateSettings` delegates to `provider.setValue` rather than `contextProxy.setValue` so the view-local buffer/pin sync path (`_saveViewLocalStateFromMutation`) runs for settings edits too.

## Measurements

`git diff --numstat 43b52aa11` (stack base, F1b head) — a+d total: 545 (528 insertions, 17 deletions):

| file | + | − |
| --- | --- | --- |
| `src/core/webview/webviewMessageHandler.ts` | 32 | 8 |
| `src/core/webview/__tests__/webviewMessageHandler.spec.ts` | 117 | 1 |
| `webview-ui/src/utils/vscode.ts` | 62 | 7 |
| `webview-ui/src/utils/__tests__/vscode.spec.ts` (new) | 216 | 0 |
| `webview-ui/src/context/ExtensionStateContext.tsx` | 4 | 1 |
| `webview-ui/src/context/__tests__/ExtensionStateContext.spec.tsx` | 97 | 0 |

Composition note: 430 of the 528 inserted lines are tests (spec files 117 + 97 + 216); production additions are 98 lines. a+d is above the 400 soft budget because the unit ships both webview-side and extension-side behavior with unit + integration tests at each layer; it is well under the 1000 hard cap.

Changed executable lines (stryker-diff): 65 (32 extension + 33 webview changed lines) — cap ≤500.
Raw mutants (stryker-diff): 65 — cap ≤400.

## Gates

- eslint: `eslint --prune-suppressions --max-warnings=0` exit 0 per touched file (src: `webviewMessageHandler.ts` + spec; webview-ui: `vscode.ts`, `vscode.spec.ts`, `ExtensionStateContext.tsx` + spec). `src/eslint-suppressions.json` unchanged — suppression counts did not increase (a prune run that only re-indented the file with zero count change was reverted).
- check-types: `src` exit 0; `webview-ui` exit 0.
- vitest: `webviewMessageHandler.spec.ts` 85/85 pass (81 base + 4 new launch tests); `vscode.spec.ts` 9/9 (new spec); `ExtensionStateContext.spec.tsx` 24/24 (21 base + 3 new). `ClineProvider.spec.ts` not affected (no `ClineProvider.ts` changes in this unit).
- prettier: `--check` exit 0 on all six touched files (CRLF checkout normalized via `--write`).
- stryker-diff (base `43b52aa11`, head `090d2c87e`): 65 raw mutants — 59 Killed, 0 Survived, 0 NoCoverage (6 Ignored equivalent mutants, the documented CS `Stryker disable` comments in `vscode.ts` L91/L122). Caps: 0 Survived / 0 NoCoverage in changed code, 65 changed executable lines ≤ 500, 65 raw mutants ≤ 400.

## Parked / documented

Observed in the CS diff but not ported (register items, to be tracked by the series ledger):

- Mojibake comment hunk in the WMH diff (`—?` corruption in the `requestRouterModels` opencode-go comment): base comment `// Deliberately no opencodeGoApiKey — the endpoint is public.` kept as-is.
- Unused `defaultModeSlug` import in the WMH spec: not ported (F3 re-adds it with its use).
- kimi-code OAuth `try/catch` hunk in `requestRouterModels` + its `webviewMessageHandler.routerModels.spec.ts` additions: not ported (review-hardening hunk outside F1c scope).
- `ApiConfigManager.tsx` className tweak: not ported (not part of the F1c row).
- `ApiConfigManager.visual.tsx` deletion + screenshot baselines: not ported (visual-suite churn outside F1c scope).
- providers/*, types, fetchers, e2e fixtures, and repo-config churn (`.coderabbit.yaml`, `label-pr-review-state.yml`, `.gitignore`, `CONTRIBUTING.md`, `ClineProvider.ts` changes, parallel-mode/sticky-mode specs, etc.): not ported (belong to the other series units).

## Porting notes

Hand-ported from CS commit `e9a44b2fa` (base of record `0d937c050`), hunk by hunk; no cherry-pick.

Ported:

- `src/core/webview/webviewMessageHandler.ts`: `webviewDidLaunch` handler — `await provider.setViewStateId(message.viewStateId)`; launch-time re-pin block (validate merged view-local name, then shared global, re-pin the view via `provider.saveViewState` when the global is still valid, else legacy global repair); `updateSettings` routed through `provider.setValue`.
- `webview-ui/src/utils/vscode.ts`: `VSCodeAPIWrapper` fallback state, `createViewStateId`/`getViewStateId`, browser-fallback `getState`/`setState` with the CS `Stryker disable` comments.
- `webview-ui/src/context/ExtensionStateContext.tsx`: launch effect posts `webviewDidLaunch` with `viewStateId`.
- `webview-ui/src/utils/__tests__/vscode.spec.ts`: new spec, all 9 tests (id reuse, create+persist, in-memory fallback, id shape, stored-state edge cases).
- `src/core/webview/__tests__/webviewMessageHandler.spec.ts`: `RooCodeSettings` import, `saveViewState` mock, `setValue` mock delegating to `contextProxy.setValue`, and the `webviewDidLaunch` describe (CS verbatim, plus one CS deviation below).

CS deviation (mutation coverage): the CS launch tests as-is leave 2 mutants alive in the re-pin block — the `StringLiteral` on the `getGlobalState("currentApiConfigName")` key (the CS mock `getValue` returns the canned value for any key, so a mutated key is unobservable) and the `ConditionalExpression` on `if (name)` (every CS test leaves `name` truthy). To satisfy the 0-survived stryker-diff gate without an escape hatch, the `getValue` mock in the launch describe is key-aware (`"currentApiConfigName"` → `"shared-profile"`, anything else → `undefined`), and one additional test covers the falsy-`name` legacy repair (selection recorded, no profile activation). This matches the CS commit's own "harden viewStateId mutation coverage" intent.
- `webview-ui/src/context/__tests__/ExtensionStateContext.spec.tsx`: `@src/utils/vscode` mock, `ViewLocalStateTestComponent`, and its 3 tests (launch post with/without id; view-local reseed contract).

Not ported (per the register above): the six parked items — verified by full-file diff against the CS final state: every ported file is byte-identical to the CS tree (modulo the intentionally skipped hunks).

### CODE REVIEW

**Comment by easonLiangWorldedtech** [comment]
## CI status update (2026-09-14)

Merged the latest main (`ba46d1f34`) to advance the merge-base. The previous `mutation-diff` failure was base drift, not a defect in this PR: the gate's diff (merge-base → PR merge-ref) included ~a week of main's own changed lines, and main's code is not covered by this PR's tests. Upstream #1610 has also made surviving mutants advisory since.

The diff now contains only this PR's own changes; expecting a green re-run of the mutation gate on the new head.

**Comment by easonLiangWorldedtech** [comment]
## Correction to the status comment above (2026-09-14)

The re-run on head `c822ede4a` failed at the executable-line scope step:

> Mutation gate failed: extension has 605 changed executable lines (limit 500).

Same stacked-series effect as #1550: this PR's own layer is ~72 a+d; the excess comes from the layers below it in the stack (notably #1546 f1a at ~900 a+d). Once #1546 lands on main, this PR's delta falls under the 500-line cap and the gate will pass on the next main merge. No code change is expected from this PR.

File: src/core/webview/webviewMessageHandler.ts [hunk 1]
```diff
@@ -579,7 +579,9 @@ export const webviewMessageHandler = async (
 				provider.resolveWebviewThemeFixtureProbe(message.requestId, message.themeFixture)
 			}
 			break
-		case "webviewDidLaunch":
+		case "webviewDidLaunch": {
+			await provider.setViewStateId(message.viewStateId)
```
[the remark this answers is not in the document]
Comment (easonLiangWorldedtech) [reply src/core/webview/webviewMessageHandler.ts]: Addressed in ac6bcd5cbf (this unit's own layer): the webviewDidLaunch flow wraps provider.setViewStateId in try/catch — a failed durable view-state registration is logged and the rest of the launch (initial state, theme, API configuration, launch-state setup) continues, since the view id is an optimization while postStateToWebview and isViewLaunched are required for the webview to function.

File: src/core/webview/__tests__/webviewMessageHandler.spec.ts [hunk 2]
```diff
@@ -2104,6 +2247,23 @@ describe("webviewMessageHandler - telemetrySetting", () => {
...
+			getMcpHub: unknown
+			providerSettingsManager: unknown
+			getStateToPostToWebview: unknown
+		}
+		return {
+			getMcpHub: view.getMcpHub,
+			providerSettingsManager: view.providerSettingsManager,
+			getStateToPostToWebview: view.getStateToPostToWebview,
+		}
+	})()
```
[the remark this answers is not in the document]
Comment (easonLiangWorldedtech) [reply src/core/webview/__tests__/webviewMessageHandler.spec.ts:2265]: Addressed in 31a1d80c4 (F1b layer of this stack, included in this PR's cumulative diff): the launch-suite snapshot now lives at module scope and the webviewDidLaunch suite restores the mockClineProvider members it replaces in its own afterEach, so the launch doubles cannot leak into the tests between the suites.

File: src/core/webview/ClineProvider.ts [hunk 3]
```diff
@@ -1736,6 +2079,14 @@ export class ClineProvider
 	): Promise<void> {
 		const task = targetTask
 
+		// A cancelled or timed-out switch must not be partially applied: bail out
+		// before the task history / _taskMode writes as well as the durable mode
+		// write below. The pre-write check further down still covers aborts that
+		// land while the task writes are in flight.
+		if (signal?.aborted) {
+			return
+		}
```
[the remark this answers is not in the document]
Comment (easonLiangWorldedtech) [reply src/core/webview/ClineProvider.ts:2089]: Addressed in c94b42781 (F1a layer of this stack, now merged through F1b into this unit): the in-flight window is now closed. If the mutation signal aborts while the task-history write is still pending, handleModeSwitchUnlocked rolls back the landed history write (re-writing the pre-switch item) and returns before the TaskModeSwitched emission and before the _taskMode assignment, so the task instance, task history, and provider state cannot retain the new mode. TaskModeSwitched is emitted only after persistence has settled and the signal is still un-aborted, and a view with no history item aborts before any write as well. The deterministic cancellation coverage requested is ClineProvider.sticky-mode.spec.ts, "rolls back the landed history write and leaves no partial mode state when the signal aborts in flight": it defers a controlled updateTaskHistory() call, aborts while it is pending, and asserts the rollback restores the original item, _taskMode is unchanged, TaskModeSwitched is not emitted, and no durable mode write occurs.

File: src/core/webview/ClineProvider.ts [hunk 4]
```diff
@@ -1956,13 +2349,73 @@ export class ClineProvider
...
+					error instanceof Error ? error.message : String(error)
+				}`,
+			)
+		}
+
+		// Capture this view's pin before setValue rewrites it: a view pinned to the
+		// deleted profile while the global selection points elsewhere must still be
+		// reconfigured, or getState() would keep the deleted profile's settings under
+		// the surviving profile's name.
+		const viewWasPinnedToDeleted = this.viewLocalState.currentApiConfigName === profileToDelete.name
```
[the remark this answers is not in the document]
Comment (easonLiangWorldedtech) [reply src/core/webview/ClineProvider.ts:2413]: Addressed in c94b42781 (F1a layer of this stack, now merged through F1b into this unit): deleteProviderProfile now captures viewWasPinnedToDeleted and deletedWasGlobal before any rewrite and preserves the local pin unless it names the deleted profile. setValue("currentApiConfigName", profileToActivate) — which updates both the view-local buffer and the shared store — runs only when this view is pinned to the deleted profile; when the deleted profile was the global selection but the view is pinned elsewhere, only the shared store is updated via contextProxy.setValue. The existing apiConfiguration reconfiguration guard is unchanged, and the test now expects the local keeper-profile pin to survive: getValues() reports the local pin while the shared store has moved to the surviving selection.

File: src/core/webview/ClineProvider.ts [hunk 5]
```diff
@@ -1956,13 +2349,73 @@ export class ClineProvider
...
+
+		if (
+			(profileToDelete.name === globalSettings.currentApiConfigName || viewWasPinnedToDeleted) &&
+			survivingSettings
+		) {
+			// The deleted profile was the active one (globally, or for this view), so
+			// the shared provider keys and this view's buffer still carry its settings;
+			// replace both so getState() reports the surviving profile's configuration.
+			await this.contextProxy.setProviderSettings(survivingSettings)
+			await this._saveViewLocalStateFromMutation(survivingSettings)
```
[the remark this answers is not in the document]
Comment (easonLiangWorldedtech) [reply src/core/webview/ClineProvider.ts]: Addressed in c94b42781 (F1a layer of this stack, now merged through F1b into this unit): when viewWasPinnedToDeleted is true and a surviving profile's settings resolve, the deletion path now saves { apiConfiguration: survivingSettings } into this view's local buffer via _saveViewLocalStateFromMutation, replacing the nested overlay wholesale — so getState() no longer serves the deleted profile's settings on top of the shared provider settings, and the buffer no longer holds the deleted configuration seeded by loadViewState(). A view whose local pin names another profile is not touched: its configuration is left as-is rather than being overwritten with the globally selected profile. The keeper test seeds the deleted nested configuration and asserts the surviving configuration replaces it, and asserts that a different local pin (with its nested configuration) remains unchanged.
